Show local mac when scanning wifi

cvwillegen opened the pull request:
# What does this implement/fix?

When connecting to an AP, and the AP has 'restrict connections to known MAC addresses' in the configuration, you can usually add new devices by MAC. You need to know the MAC of the ESP chip before connecting to the network.

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [X] ESP8266

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

cvwillegen committed:
Prettier date/time display after time sync

Make the date/time display after time sync prettier by always formatting into fixed width fields.

cvwillegen committed:
Log date/time using fixed width field specifiers

cvwillegen committed:
Use suggestions by CI clang-format

cvwillegen committed:
Use suggestions by CI clang-format

cvwillegen committed:
Merge pull request #1 from cvwillegen/prettier-date-time-display-after-time-sync

Prettier date/time display after time sync

cvwillegen committed:
Merge branch 'esphome:dev' into dev

cvwillegen committed:
Merge branch 'esphome:dev' into dev

cvwillegen committed:
Merge branch 'esphome:dev' into dev

cvwillegen committed:
Merge branch 'esphome:dev' into dev

cvwillegen committed:
Show local MAC when scanning for the first time

When starting WiFi scan for the first time, the local MAC is not shown in the logs. When an AP has 'restrict adding new wireless devices' in its config (only connect to known MAC addresses), the ESP can't connect. You need to add the local MAC into the AP configuration, so you need this info.

RoboMagus reviewed (commented):
This does not seem to be the right place for logging the local MAC, as it has nothing to do with the WiFi scan itself.
The local mac is already logged as part of the config dump when connecting to the logger. that should be sufficient.

cvwillegen commented:
I connected a (new) Atom echo to my RPI running HA.

I took the config from ``https://github.com/esphome/media-players/blob/main/m5stack-atom-echo.yaml``.

I added
```
wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
```

Then, I clicked ``Install``. Clicked on ``Plug into the computer running ESPHome Dashboard`` and selected the correct serial port (``M5Stack``).

After quite some time, I see:
``Serial port /dev/ttyUSB2
Connecting......
Chip is ESP32-PICO-D4 (revision 1)
Features: WiFi, BT, Dual Core, 240MHz, Embedded Flash, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 4c:75:25:a2:81:fc
Uploading stub...
``

I think this is the correct MAC address.

Some time after that, I get:

```
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.1 seconds (effective 584.7 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
INFO Successfully uploaded program.
INFO Starting log output from /dev/ttyUSB2 with baud rate 115200
[19:27:19]\xff[I][logger:243]: Log initialized
[19:27:19][C][ota:461]: There have been 0 suspected unsuccessful boot attempts.
[19:27:19][D][esp32.preferences:114]: Saving preferences to flash...
[19:27:19][I][app:029]: Running through setup()...
[19:27:19][D][binary_sensor:034]: 'M5Stack Atom Echo Button': Sending initial state OFF
[19:27:19][C][fastled:012]: Setting up FastLED light...
[19:27:19][C][light:036]: Setting up light 'M5Stack Atom Echo'...
[19:27:19][D][light:035]: 'M5Stack Atom Echo' Setting:
[19:27:19][D][light:040]:   Color mode: RGB
[19:27:19][D][light:050]:   Brightness: 100%
[19:27:19][D][light:054]:   Color brightness: 100%
[19:27:19][D][light:058]:   Red: 100%, Green: 100%, Blue: 100%
[19:27:19][C][wifi:037]: Setting up WiFi...
[19:27:19][D][wifi:373]: Starting scan...
[19:27:19][C][wifi:374]:   Local MAC: 4C:75:25:A2:81:FC
[19:27:19][D][esp32.preferences:114]: Saving preferences to flash...
[19:27:21][D][wifi:389]: Found networks:
[19:27:21][I][wifi:433]: - [redacted] [redacted]▂▄▆█
[19:27:21][D][wifi:434]:     Channel: 1
[19:27:21][D][wifi:435]:     RSSI: -60 dB
[19:27:21][I][wifi:433]: - [redacted] [redacted]▂▄▆█
[19:27:21][D][wifi:434]:     Channel: 11
[19:27:21][D][wifi:435]:     RSSI: -86 dB
[19:27:21][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:27:21][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:27:21][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:27:21][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:27:21][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:27:21][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:27:21][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:27:21][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:27:21][I][wifi:248]: WiFi Connecting to [redacted]...
[19:27:22][W][wifi_esp32:495]: Event: Disconnected ssid=[redacted] bssid=[redacted] reason='Authentication Failed'
[19:27:22][W][wifi:553]: Connecting to WiFi network failed. Are the credentials wrong?
[19:27:22][I][wifi:248]: WiFi Connecting to [redacted]...
[19:27:22][W][wifi:553]: Connecting to WiFi network failed. Are the credentials wrong?
[19:27:22][W][wifi_esp32:495]: Event: Disconnected ssid=[redacted] bssid=[redacted] reason='Authentication Failed'
```

Note: the ``[19:27:19][C][wifi:374]:   Local MAC: 4C:75:25:A2:81:FC`` shown in the log is the MAC address logging that I added in this PR.

If I close this window, and click on ``Logs`` for the newly added device, I get:

```
INFO Reading configuration /config/esphome/atom-echo-2.yaml...
INFO Starting log output from /dev/ttyUSB2 with baud rate 115200
[19:30:29][D][wifi:389]: Found networks:
[19:30:29][I][wifi:433]: - [redacted] [redacted]▂▄▆█
[19:30:29][D][wifi:434]:     Channel: 11
[19:30:29][D][wifi:435]:     RSSI: -84 dB
[19:30:29][I][wifi:433]: - [redacted] [redacted]▂▄▆█
[19:30:29][D][wifi:434]:     Channel: 1
[19:30:29][D][wifi:435]:     RSSI: -60 dB
[19:30:29][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:30:29][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:30:29][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:30:29][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:30:29][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:30:29][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:30:29][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:30:29][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:30:29][D][wifi:438]: - [redacted] [redacted]▂▄▆█
[19:30:29][I][wifi:248]: WiFi Connecting to [redacted]...
[19:30:29][W][wifi:553]: Connecting to WiFi network failed. Are the credentials wrong?
[19:30:29][W][wifi_esp32:495]: Event: Disconnected ssid=[redacted] bssid=[redacted] reason='Authentication Failed'
[19:30:29][I][wifi:248]: WiFi Connecting to [redacted]...
[19:30:29][W][wifi:553]: Connecting to WiFi network failed. Are the credentials wrong?
[19:30:29][W][wifi_esp32:495]: Event: Disconnected ssid=[redacted] bssid=[redacted] reason='Authentication Failed'
[19:30:34][D][wifi:373]: Starting scan...
[19:30:34][C][wifi:374]:   Local MAC: 4C:75:25:A2:81:FC
```
Again, the ``[19:30:34][C][wifi:374]:   Local MAC: 4C:75:25:A2:81:FC`` is added by this PR.

I need to add the WiFi MAC to my router so that the ESP can connect to it, but in this log I am unable to see the MAC. The only place I can see it is during the flashing process, and if you close that window, the information is lost.

Can you suggest a more appropriate place to log the MAC address, so that I can add it to my router's allow list?

RoboMagus commented:
Upon connecting the logger to an ESPHome device, the first thing I get presented is the config as shown below.
Local MAC is already logged on the first line under the WiFi section.

```
[19:50:17][C][wifi:491]: WiFi:
[19:50:17][C][wifi:353]:   Local MAC: 01:23:45:67:89:AB
[19:50:17][C][wifi:354]:   SSID: [redacted]
[19:50:17][C][wifi:355]:   IP Address: 192.168.1.220
[19:50:17][C][wifi:356]:   BSSID: [redacted]
[19:50:17][C][wifi:358]:   Hostname: 'esphome-dummy'
[19:50:17][C][wifi:360]:   Signal strength: -77 dB ▂▄▆█
[19:50:17][C][wifi:364]:   Channel: 1
[19:50:17][C][wifi:365]:   Subnet: 255.255.255.0
[19:50:17][C][wifi:366]:   Gateway: 192.168.1.1
[19:50:17][C][wifi:367]:   DNS1: 0.0.0.0
[19:50:17][C][wifi:368]:   DNS2: 0.0.0.0
[19:50:17][C][logger:275]: Logger:
[19:50:17][C][logger:276]:   Level: DEBUG
[19:50:17][C][logger:277]:   Log Baud Rate: 115200
[19:50:17][C][logger:278]:   Hardware UART: UART0
```

If you don't manually set the global loglevel, it defaults to `debug`, which is needed to get the information you want.

cvwillegen commented:
This is when it's connected via WiFi, isn't it? Mine hasn't been connected, since it can't (yet), because I need the MAC for that! If you connect it to a USB cable, what do you get in the log output?

RoboMagus commented:
@cvwillegen, you are indeed correct that this information is only printed after WiFi is connected. My appologies, I rarely end up with a case wher I do not yet have the device connected to my network.

However, as the config logging (that includes the Local MAC) is exposed through the `dump_config` function, it is quite easy to create a config that prints it out on startup:

Logging:
```
[I][logger:243]: Log initialized
[C][ota:461]: There have been 3 suspected unsuccessful boot attempts.
[I][app:029]: Running through setup()...
[C][wifi:491]: WiFi:
[C][wifi:353]:   Local MAC: 01:23:45:67:89:AB
[C][wifi:354]:   SSID: [redacted]
[C][wifi:355]:   IP Address: 0.0.0.0
[C][wifi:356]:   BSSID: 50:E0:39:AE:24:91
[C][wifi:358]:   Hostname: 'wifi_config_test'
[C][wifi:360]:   Signal strength: 31 dB ▂▄▆█
[C][wifi:364]:   Channel: 1
[C][wifi:365]:   Subnet: 0.0.0.0
[C][wifi:366]:   Gateway: 0.0.0.0
[C][wifi:367]:   DNS1: 0.0.0.0
[C][wifi:368]:   DNS2: 0.0.0.0
[C][wifi:037]: Setting up WiFi...
[I][wifi:248]: WiFi Connecting to [redacted]...
[W][wifi_esp8266:482]: Event: Disconnected ssid=[redacted] bssid=[redacted] reason='4-Way Handshake Timeout'
[W][wifi:536]: Error while connecting to network.
[W][wifi:572]: Restarting WiFi adapter...
[W][wifi_esp8266:482]: Event: Disconnected ssid=[redacted] bssid=[redacted] reason='Authentication Failed'
[I][wifi:248]: WiFi Connecting to [redacted]...
[W][wifi_esp8266:482]: Event: Disconnected ssid=[redacted] bssid=[redacted] reason='4-Way Handshake Timeout'
[W][wifi:536]: Error while connecting to network.
[W][wifi:572]: Restarting WiFi adapter...
[W][wifi_esp8266:482]: Event: Disconnected ssid=[redacted] bssid=[redacted] reason='Authentication Failed'
```

Config:
```yaml
esphome:
  name: wifi_config_test

  on_boot:
    then:
      - lambda: "id(wifiComponent)->dump_config();"

wifi:
  id: wifiComponent
  ssid: !secret wifi_ssid
  password: !secret wifi_password

logger:
  level: DEBUG
```

I hope this information helps you out.

cvwillegen commented:
So, you're turning down a one line PR, and instead offer a totally unrelated config that people need to flash into their device in order to be able to use them on their network... And how are you going to tell people about this solution? Put it on the Wiki, or should people go look for it themselves?

Well, it _is_ your decision, in the end.

I'm sorry if I come over a bit harsh, I realise how I sound, but it strikes me as very odd that you can solve a problem with one line, and decline it for everyone...

cvwillegen commented:
Also, please check https://github.com/esphome/esphome/pull/2428 to see where the information you mentioned originally came from. It was already missing some time ago...

RoboMagus commented:
> So, you're turning down a one line PR, ...

I'm not turning down anything here as I do not have the rights to decline or approve a merge request. Just offering my insights and oppinion. However as a general rule of thumb, adding code to places where it does not belong to solve for a niche issue that a small minority of ESPHome users might run into is just bad practice. Especially since there are already methods in place that allow end users to mittigate that issue.

I do agree with you on the availability on that knowledge though. I too am not a big fan of how one would have to find out on how to create a config to deal with this.
But I would propose that _that_ is the core issue here and _that_ is what should be foccussed on, instead of implementing the change you proposed.

RoboMagus commented:
Alternatively, I would propose adding a call to `dump_config` at the beginning of the `WiFiComponent::setup()` such that the config is printed before the device enters the failing connect loop.

```c++
void WiFiComponent::setup() {
  ESP_LOGCONFIG(TAG, "Setting up WiFi...");
  this->dump_config(); // Print config before connecting
  this->last_connected_ = millis();
  this->wifi_pre_setup_();

  ...
```

cvwillegen committed:
Reflect advise from reviewer

Instead of just dumping the MAC address when starting a WiFi scan, dump the WiFi config at the right moment in the setup.

cvwillegen commented:
Yes, that was a good idea! I'm sorry that I misunderstood your advise at first.

I _did_ notice one small snag: When you close the Log window and re-open it, it doesn't show the WiFi info, since it's only shown after boot... but I can live with that.

cvwillegen committed:
Fix comment indent

jesserockz commented on esphome/components/wifi/wifi_component.cpp:
```diff
@@ -35,6 +35,7 @@ float WiFiComponent::get_setup_priority() const { return setup_priority::WIFI; }
 
 void WiFiComponent::setup() {
   ESP_LOGCONFIG(TAG, "Setting up WiFi...");
+  this->dump_config();  // Print config before connecting
```
The dump config is really only useful once it's already connected, let's just log the MAC only...

```suggestion
  ESP_LOGCONFIG(TAG, "  Local MAC: %s", get_mac_address_pretty().c_str());
```

jesserockz committed:
Update esphome/components/wifi/wifi_component.cpp

jesserockz commented on esphome/components/wifi/wifi_component.cpp:
```diff
@@ -35,6 +35,7 @@ float WiFiComponent::get_setup_priority() const { return setup_priority::WIFI; }
 
 void WiFiComponent::setup() {
   ESP_LOGCONFIG(TAG, "Setting up WiFi...");
+  ESP_LOGCONFIG(TAG, "  Local MAC: %s", get_mac_address_pretty().c_str());
```
```suggestion
  ESP_LOGCONFIG(TAG, "  Local MAC: %s", get_mac_address_pretty().c_str());
```

jesserockz committed:
Update esphome/components/wifi/wifi_component.cpp
